azureeyemodule/app/util/time_aligned_buffer.cpp

```diff
@@ -0,0 +1,150 @@
+    for (const auto &tup : this->timestamped_frames)
+    {
+        auto tup_ts = std::get<1>(tup);
+        if (tup_ts < oldest_ts)
+        {
+            std::tie(oldest_frame, oldest_ts) = tup;
+        }
+
+        // Also find the best match
+        auto timedelta = (tup_ts > timestamp) ? (tup_ts - timestamp) : (timestamp - tup_ts);
+        if (timedelta < smallest_time_difference)
+        {
+           best_match_ts = std::get<1>(tup);
+           smallest_time_difference = timedelta;
+        }
+    }
+
+    // Sanity check
+    assert(oldest_ts != LLONG_MAX);
+    assert(best_match_ts != 0);
+    assert(smallest_time_difference != LLONG_MAX);
+    #ifdef DEBUG_TIME_ALIGNMENT
+        util::log_debug("New Inference: Matched " + util::timestamp_to_string(timestamp) + " with " + util::timestamp_to_string(best_match_ts));
+    #endif
+
+    if (oldest_ts > timestamp)
+    {
+        #ifdef DEBUG_TIME_ALIGNMENT
+            util::log_debug("New Inference: oldest frame is not old enough! We will store " + std::to_string(this->n_timestamped_frames * 2) + " frames now.");
+        #endif
+
+        // If oldest frame occurs after this
+        // frame's timestamp, we are overwriting our buffer frames before the network can inference even a single time.
+        // Therefore the buffer is not large enough and needs to be resized.
+        this->n_timestamped_frames *= 2;
+
+        // Use this frame as the best one, but don't remove it, since it doesn't really match.
+        assert(best_match_ts == oldest_ts);
+        return std::vector<cv::Mat>{oldest_frame};
+    }
+    else
+    {
+        // If the oldest frame occurs before this inference, then the frame we are inferencing on
+        // is somewhere in our buffer. Search through to find it and all older frames.
+        std::vector<cv::Mat> best_and_older;
+        std::vector<size_t> indices_to_erase;
+        for (size_t i = 0; i < this->timestamped_frames.size(); i++)
+        {
+            const auto &tup = this->timestamped_frames.at(i);
+            if (std::get<1>(tup) <= best_match_ts)
+            {
+                indices_to_erase.push_back(i);
+                best_and_older.push_back(std::get<0>(tup));
+            }
+        }
+
+        // Update the default value
+        this->default_value = best_and_older.back();
+
+        // Remove all the old frames.
+        assert(best_and_older.size() > 0);
+        assert(indices_to_erase.size() > 0);
+        auto start = indices_to_erase.at(0);
+        auto end = indices_to_erase.back() + 1;
+        this->timestamped_frames.erase(this->timestamped_frames.begin() + start, this->timestamped_frames.begin() + end);
+
+        // We need to decrement index by an amount equal
+        // to the number of indices we removed that are less than it.
+        size_t amount = 0;
+        for (auto i : indices_to_erase)
+        {
+            if (i < this->index)
```

Comment:
First we figure out what items we need to erase (all frames that are older than or equal to the best match).

Then we actually erase them from the vector.

But because we are a circular buffer wrapper around a vector, we need to make sure that the write index gets updated to be valid, in case we removed items from the vector (thereby shrinking the vector).

For example, say we found that the indices that need removing are [0, 1, 2, 8, 9] (because it is circular, and 8 and 9 are the last items in the buffer, and the oldest). In this case, the write index is probably at 8. Removing 8 and 9 doesn't matter for adding new frames at the write index, but removing 0, 1, and 2 does. Removing those frames means we need to decrement our write index by 3, since there are 3 fewer items in the vector:

```
                      Write Index
                          |
[ 0 1 2 3 4 5 6 7 8 9]
  ^^^ Erasing these three items leads to:

[ 0 1 2 3 4 5 6 ] But my write index still points to 8, which is off the end of the array.
```